traitsui/qt4/datetime_editor.py

```diff
@@ -15,60 +15,79 @@
 """
 
 
-import datetime
-
-from pyface.qt import QtGui
+from pyface.qt import QtGui, qt_api
 from pyface.qt.QtCore import QDateTime
-from traits.api import Instance
+from traits.api import Datetime, on_trait_change
 
 from .editor import Editor
 from .editor_factory import ReadonlyEditor as BaseReadonlyEditor
 
 
+
 class SimpleEditor(Editor):
     """ Simple Traits UI time editor that wraps QDateTimeEdit.
     """
 
     #: the earliest datetime allowed by the editor
-    minimum_datetime = Instance(datetime.datetime)
+    minimum_datetime = Datetime
 
     #: the latest datetime allowed by the editor
-    maximum_datetime = Instance(datetime.datetime)
+    maximum_datetime = Datetime
 
     def init(self, parent):
         """ Finishes initializing the editor by creating the underlying toolkit
             widget.
         """
-        self.control = QtGui.QDateTimeEdit()
+        # set min and max early, don't wait for editor sync
         self.minimum_datetime = self.factory.minimum_datetime
         self.maximum_datetime = self.factory.maximum_datetime
-        self.update_date_range()
+
+        self.control = QtGui.QDateTimeEdit()
         self.control.dateTimeChanged.connect(self.update_object)
+        self.update_date_range()
 
     def update_editor(self):
         """ Updates the editor when the object trait changes externally to the
             editor.
         """
         value = self.value
-        if value:
-            q_datetime = QDateTime(value)
-            self.control.setTime(q_datetime)
+        if self.value:
+            if self.minimum_datetime and self.minimum_datetime > value:
+                value = self.minimum_datetime
+            elif self.maximum_datetime and self.value > self.maximum_datetime:
+                value = self.maximum_datetime
+            try:
+                q_datetime = QDateTime(value)
+            except Exception:
+                pass
+            self.control.setDateTime(q_datetime)
 
     def update_object(self, q_datetime):
         """ Handles the user entering input data in the edit control.
         """
-        self.value = q_datetime.toPyDateTime()
-
+        try:
+            if qt_api == 'pyside2':
+                self.value = q_datetime.toPython()
+            else:
+                self.value = q_datetime.toPyDateTime()
+        except ValueError:
+            pass
```

Comment:
No, we want LBYL here, because there are only two values - 'pyside2' and 'pyqt5', and they are equally likely (and exceptions are expensive).

It would be useful to factor this sort of stuff out, but it probably belongs in Pyface.

Reply:
+1 to an `is_pyside` or `is_pyqt` in Pyface (or however we want to spell that), or perhaps some kind of enumeration (anticipating possible future different Python bindings for Qt)

> and exceptions are expensive

Hmm; I think any expense is insigificant at this level, and in any case this is not on a critical path. The expense or not of exceptions should not be a factor in deciding how to write this code.

Reply:
Agreed Pyface is its rightful place.
Nit: What do you think about checking `if qt_api.startswith("pyface")`? It is kind of similar to checking `sys.platform.startswith("win")` recommended by Python doc.

Reply:
In case it gets forgotten, there was also this question:

> What would cause this ValueError?

(I probably should have commented these on different lines instead of combining them, sorry about that)

Reply:
Sorry - that's the same issue as converting in the other direction: the maximum QDateTime value is greater than the maximum Python datetime.

Reply:
I see. Thank you!
(Now thinking whether we want to enforce `maximumDateTime` to be equal or less than the maximum supported by Python...)

Reply:
This is interesting... I could not run into the `ValueError`. It seems that both PyQt5 and PySide2 return a filled value, but they return *different* filled value :(

I removed the try-except-ValueError and tried this test:
```
    def test_datetime_editor_qt_datetime_out_of_bound(self):
        # The maximum datetime supported by Qt is larger than the maximum
        # datetime supported by Python
        editor_factory = DatetimeEditor()
        view = get_date_time_simple_view(editor_factory)
        init_datetime = datetime.datetime(1900, 1, 1)
        instance = InstanceWithDatetime(date_time=init_datetime)
        with self.launch_editor(instance, view) as editor:
            # the user set the datetime on the Qt widget to a value
            # too large for Python
            from pyface.qt.QtCore import QDateTime, QDate
            q_datetime = QDateTime(QDate(datetime.MAXYEAR + 1, 1, 1))
            editor.control.setDateTime(q_datetime)

            # Get the displayed value back.
            displayed_value = to_datetime(editor.control.dateTime())

        expected = datetime.datetime(
            datetime.MAXYEAR, 12, 31, 23, 59, 59, 999000)
        self.assertEqual(displayed_value, expected)
        self.assertEqual(instance.date_time, expected)
```
It passes on PySide2 but fails on PyQt5 because the filled value is different. On PyQt5, this is how it fails:
```
======================================================================
FAIL: test_datetime_editor_qt_datetime_out_of_bound (traitsui.tests.editors.test_datetime_editor.TestDatetimeEditorQt)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/traitsui/traitsui/tests/editors/test_datetime_editor.py", line 145, in test_datetime_editor_qt_datetime_out_of_bound
    self.assertEqual(displayed_value, expected)
AssertionError: datetime.datetime(7999, 12, 31, 23, 59, 59, 999000) != datetime.datetime(9999, 12, 31, 23, 59, 59, 999000)
```

Reply:
It seems that removing the try-except is good enough here. In the test we can't check the exact filled value and will have to comment on the fact that different backend returns different values.
As long as the value on trait is consistent with the view, I think we are good.